google/cloud/internal/oauth2_external_account_credentials_test.cc

```diff
@@ -0,0 +1,492 @@
+  auto const test_url = std::string{"https://sts.example.com/"};
+  auto const expected_access_token = std::string{"test-access-token"};
+  auto const expected_expires_in = std::chrono::seconds(3456);
+  auto const json_response = nlohmann::json{
+      {"access_token", expected_access_token},
+      {"invalid-expires_in", expected_expires_in.count()},
+      {"issued_token_type", "urn:ietf:params:oauth:token-type:access_token"},
+      {"token_type", "Bearer"},
+  };
+  auto mock_source = [](Options const&) {
+    return make_status_or(internal::SubjectToken{"test-subject-token"});
+  };
+  auto const info = ExternalAccountInfo{
+      "test-audience",
+      "test-subject-token-type",
+      test_url,
+      mock_source,
+  };
+  MockClientFactory client_factory;
+  EXPECT_CALL(client_factory, Call).WillOnce([&]() {
+    auto mock = absl::make_unique<MockRestClient>();
+    EXPECT_CALL(*mock, Post(_, An<FormDataType const&>()))
+        .WillOnce([test_url, info, response = json_response.dump()](
+                      auto const&, auto const&) {
+          return MakeMockResponseSuccess(response);
+        });
+    return mock;
+  });
+
+  auto credentials =
+      ExternalAccountCredentials(info, client_factory.AsStdFunction());
+  auto const now = std::chrono::system_clock::now();
+  auto access_token = credentials.GetToken(now);
+  EXPECT_THAT(access_token, StatusIs(StatusCode::kInvalidArgument));
+  EXPECT_THAT(access_token.status().error_info().domain(), "gcloud-cpp");
+}
+
+TEST(ExternalAccount, InvalidExpiredIn) {
+  auto const test_url = std::string{"https://sts.example.com/"};
+  auto const expected_access_token = std::string{"test-access-token"};
+  auto const json_response = nlohmann::json{
+      {"access_token", expected_access_token},
+      {"expires_in", "--invalid--"},
+      {"issued_token_type", "urn:ietf:params:oauth:token-type:access_token"},
+      {"token_type", "Bearer"},
+  };
+  auto mock_source = [](Options const&) {
+    return make_status_or(internal::SubjectToken{"test-subject-token"});
+  };
+  auto const info = ExternalAccountInfo{
+      "test-audience",
+      "test-subject-token-type",
+      test_url,
+      mock_source,
+  };
+  MockClientFactory client_factory;
+  EXPECT_CALL(client_factory, Call).WillOnce([&]() {
+    auto mock = absl::make_unique<MockRestClient>();
+    EXPECT_CALL(*mock, Post(_, An<FormDataType const&>()))
+        .WillOnce([test_url, info, response = json_response.dump()](
+                      auto const&, auto const&) {
+          return MakeMockResponseSuccess(response);
+        });
+    return mock;
+  });
+
+  auto credentials =
+      ExternalAccountCredentials(info, client_factory.AsStdFunction());
+  auto const now = std::chrono::system_clock::now();
+  auto access_token = credentials.GetToken(now);
+  EXPECT_THAT(access_token, StatusIs(StatusCode::kInvalidArgument));
+  EXPECT_THAT(access_token.status().error_info().domain(), "gcloud-cpp");
+}
+
```

Comment:
what about a test for `RestPayloadError`:
https://github.com/googleapis/google-cloud-cpp/blob/002e5879568b8a35eb8dcf99788db0f04b1550f7/google/cloud/internal/oauth2_external_account_credentials.cc#L54-L55

And another one for `InvalidJson`:
https://github.com/googleapis/google-cloud-cpp/blob/002e5879568b8a35eb8dcf99788db0f04b1550f7/google/cloud/internal/oauth2_external_account_credentials.cc#L64-L69

:)

Reply:
Thanks